diff -Nur nacl-20110221.orig/cpucycles/mips.c nacl-20110221/cpucycles/mips.c
--- nacl-20110221.orig/cpucycles/mips.c	2012-07-03 10:41:37.887065263 +0200
+++ nacl-20110221/cpucycles/mips.c	2012-07-03 10:42:20.704921053 +0200
@@ -18,7 +18,7 @@
 {
   struct timeval t;
   unsigned int cc;
-  asm volatile(".byte 59; .byte 16; .byte 2; .byte 124; move %0,$2" : "=r"(cc) : : "$2");
+  asm volatile(".long 2080510011; move %0,$2" : "=r"(cc) : : "$2");
   gettimeofday(&t,(struct timezone *) 0);
   result[0] = cc;
   result[1] = t.tv_usec;
