diff -Nur djbdns-1.05.orig/query.c djbdns-1.05/query.c
--- djbdns-1.05.orig/query.c	2012-07-14 11:26:52.472055498 +0200
+++ djbdns-1.05/query.c	2012-07-14 11:28:31.971555036 +0200
@@ -705,6 +705,14 @@
       save_finish(DNS_T_PTR,t1,ttl,0);
     }
     else if (byte_equal(type,2,DNS_T_NS)) {
+      if (posauthority <= records[i]) {
+        unsigned int cachedttl;
+        dlen = dns_domain_length(t1);
+        byte_copy(key,2,DNS_T_NS);
+        byte_copy(key + 2,dlen,t1);
+        case_lowerb(key + 2,dlen);
+        if (cache_get(key,dlen + 2,&cachedlen,&cachedttl,0)) if (cachedttl < ttl) ttl = cachedttl;
+      }
       save_start();
       while (i < j) {
         pos = dns_packet_skipname(buf,len,records[i]); if (!pos) goto DIE;
